lib/CGameState.h

```diff
@@ -279,29 +279,52 @@ struct DLL_EXPORT DuelParameters
 class CPathfinder : private CGameInfoCallback
 {
 private:
-	bool allowEmbarkAndDisembark;
-	bool allowTeleportTwoWay; // Two-way monoliths and Subterranean Gate
-	bool allowTeleportOneWay; // One-way monoliths with one known exit only
-	bool allowTeleportOneWayRandom; // One-way monoliths with more than one known exit
-	bool allowTeleportWhirlpool; // Force enabled if hero protected or unaffected (have one stack of one creature)
+	struct PathfinderOptions
+	{
+		bool useFlying;
+		bool useWaterWalking;
+		bool useEmbarkAndDisembark;
+		bool useTeleportTWoWay; // Two-way monoliths and Subterranean Gate
+		bool useTeleportOneWay; // One-way monoliths with one known exit only
+		bool useTeleportOneWayRandom; // One-way monoliths with more than one known exit
+		bool useTeleportWhirlpool; // Force enabled if hero protected or unaffected (have one stack of one creature)
+
+		PathfinderOptions();
+	};
+	PathfinderOptions options;
+
 	CPathsInfo &out;
 	const CGHeroInstance *hero;
 	const std::vector<std::vector<std::vector<ui8> > > &FoW;
 
 	std::list<CGPathNode*> mq; //BFS queue -> nodes to be checked
 
+	std::vector<int3> neighbours;
 
 	int3 curPos;
 	CGPathNode *cp; //current (source) path node -> we took it from the queue
 	CGPathNode *dp; //destination node -> it's a neighbour of cp that we consider
 	const TerrainTile *ct, *dt; //tile info for both nodes
 	ui8 useEmbarkCost; //0 - usual movement; 1 - embark; 2 - disembark
-	Obj destTopVisObjID;
 
+	CGObjectInstance *sTileObj;
+	CGObjectInstance *dTileObj;
+	const CGTeleport *sTileTeleport;
+	const CGTeleport *dTileTeleport;
 

```

Comment:
consider adding (inline) setter for [sd]TileObj to ensure that [sd]TileTeleport is relevant.